### ISSUE TITLE
Xcode Instruments can not read Rust dSYM file.

### ISSUE THREAD

**Issue by y-ich** [rust-lang/rust#43564]
Hi.

In the past, adding `debug = true` into Cargo.toml seemed to enable Rust source code profiling on Xcode Instruments (on macOS).
But at present, it does not work.
Trying to add dSYM path causes the following error.

    No new libraries to symbolicate
    The specified path didn't locate a dSYM for any of the selected libraries.

I will appreciate very much if anyone let me know workarounds or Rust team resolves it kindly.
Thanks.

**Comment by djg** [comment]
I'm also running into this issue trying to profile rust in Gecko. It appears the AT_decl_file entries for C++ code have absolute pathnames while the entries for Rust have relative pathnames. Is there anyway to force rust to output the absolute path to the source code in the DWARF info?

**Comment by pnkfelix** [comment]
Visiting for wg-debugging triage

Its been so long since this issue was touched, we should have someone go and confirm that it still reproduces. I'm adding E-help-wanted to reflect that.

@rustbot label: E-help-wanted

**Comment by thomcc** [comment]
I have definitely profiled rust code in Instruments in the last few months, and it worked. I can check again later today though.

I will note that I believe it worked for me back in 2017 as well.

**Comment by petar-dambovaliev** [comment]
It would be nice to have this fully working. It is much easier to narrow down what needs to be optimized when you can see the anotated source code. Flamegraphs are just useful enough to not be completely useless.

**Comment by djeedai** [comment]
Hi- It's my first time trying, but confirming all the above, that is showing Rust source code currently doesn't work on Instruments 15.3 and that as a result seeing only the disassembly and callstack is "just useful enough to be completely useless" (I know where the problem roughly is, I can't figure out exactly what part of the function).

**Comment by djeedai** [comment]
Note: as a useful workaround, grabbed from [here](https://users.rust-lang.org/t/missing-dsym-on-macos-when-building-with-cargo/97543/7), you can go into the folder of your binary and run `dsymutil <binary>`. This produces a new `<binary>.dSYM` file which Instruments then finds automatically. `dsymutil` was installed already, I think comes alongside Xcode.

**Comment by evmar** [comment]
Just set up a new Mac and tried out Instruments, so here is some info that is correct as of May 2026:

1. Out of the box, no source.
2. Without `debug = true` in `Cargo.toml` for the profile you're looking at, `dsymutil binary` says "warning: no debug symbols in executable (-arch arm64)"
3. After `dsymutil target/release/binary` I get a `.dSYM` directory under the same path, and tools like `objdump --source` and Instruments now find the source.

**Comment by madsmtm** [comment]
See [`split-debuginfo`](https://doc.rust-lang.org/cargo/reference/profiles.html#split-debuginfo), if you want a different default there (or better docs in Cargo about how to use it with Instruments), that's probably a different issue.

**Comment by evmar** [comment]
Thanks!  I think I eventually found the split-debuginfo docs on my own, but it was only after digging through bugs like these trying to figure out why Mac debugging wasn't working.  CC @alexcrichton in case you weren't aware -- `objdump --source` and Instruments on macOS do not find source code unless you dsymutil.  I will investigate whether twiddling split-debuginfo helps, probably on https://github.com/cmyr/cargo-instruments/issues/144

**Comment by evmar** [comment]
(In a brief test it appears split-debuginfo=packed causes objdump --source to print an error about a missing path.)

**Comment by evmar** [comment]
Ok, in https://github.com/cmyr/cargo-instruments/issues/144 I tried to exhaustively check the cases.

Here is the summary:

1. In a release build with debug=true (which implies unpacked debuginfo)
   1. objdump cannot find source but Instruments can
   2. *Except* for code in other crates that hasn't been inlined.
2. In a release build with debug=true, split-debuginfo=packed 
   1. objdump errors
   2. fixes the cross-crate missing source

**Comment by evmar** [comment]
I filed #156507 about 1.ii above.  I think you can close this bug as obsolete -- the initial report talks about dSYMs but Rust no longer produces one and things mostly work, aside from that bug.

**Comment by bjorn3** [comment]
dSYMs are still produced by default by rustc. Cargo sets split-debuginfo=unpacked by default to suppress this, but you can still tell cargo to keep producing dSYMs.
